src/multiversioning.zig

```diff
@@ -0,0 +1,892 @@
+            assert(std.sort.isSorted(u32, past_init.releases, {}, std.sort.asc(u32)));
+
+            assert(past_init.offsets[0] == 0);
+            for (past_init.offsets[1..], 0..) |offset, i| {
+                const calculated_offset = blk: {
+                    var calculated_offset: u32 = 0;
+                    for (past_init.sizes[0..i]) |size| {
+                        calculated_offset += size;
+                    }
+                    break :blk calculated_offset;
+                };
+                assert(offset != 0);
+                assert(offset == calculated_offset);
+            }
+
+            for (past_init.sizes) |s| assert(s != 0);
+
+            var past_releases = PastReleases{};
+            past_releases.count = count;
+
+            stdx.copy_disjoint(.inexact, u32, &past_releases.releases, past_init.releases);
+            stdx.copy_disjoint(.inexact, u128, &past_releases.checksums, past_init.checksums);
+            stdx.copy_disjoint(.inexact, u32, &past_releases.offsets, past_init.offsets);
+            stdx.copy_disjoint(.inexact, u32, &past_releases.sizes, past_init.sizes);
+            stdx.copy_disjoint(.inexact, bool, &past_releases.visits, past_init.visits);
+
+            return past_releases;
+        }
+    };
+
+    /// Covers MultiversionHeader[@sizeOf(u128)..].
+    checksum_header: u128 = undefined,
+
+    /// The AEGIS128L checksum of the binary, if the header (`.tb_mvh`) section were zeroed out.
+    /// Used to validate that the binary itself is not corrupt. Putting this in requires a bit
+    /// of trickery:
+    /// * inject a zero `.tb_mvh` section of the correct size,
+    /// * compute the hash,
+    /// * update the section with the correct data.
+    /// Used to ensure we don't try and exec into a corrupt binary.
+    checksum_binary_without_header: u128 = 0,
+
+    /// The current release is executed differently to past releases embedded in the body, so store
+    /// it separately. See exec_latest vs exec_release.
+    current_release: u32,
+
+    /// The AEGIS128L checksum of the direct output of `zig build`, for the current_release, before
+    /// any objcopy or build magic has been performed.
+    /// Used when extracting binaries from past releases to ensure a hash chain - this will be put
+    /// in the entry in past.checksums for this release by the build script.
+    current_checksum: u128,
+
+    past: PastReleases = .{},
+    reserved: [16384]u8 = std.mem.zeroes([16384]u8),
+
+    /// Parses an instance from a slice of bytes and validates its checksum. Returns a copy.
+    pub fn init_from_bytes(bytes: *[@sizeOf(MultiversionHeader)]u8) !MultiversionHeader {
+        const self = std.mem.bytesAsValue(MultiversionHeader, bytes).*;
+
+        const checksum_calculated = self.calculate_header_checksum();
+
+        if (checksum_calculated != self.checksum_header) {
+            return error.ChecksumMismatch;
+        }
+
```

Comment:
> and that reserved is zeroed?

This I want to avoid; so that we can add information there in the future and still have it pass with previous builds.